src/data/helper/createDimensions.ts

```diff
@@ -44,33 +51,350 @@ export type CoordDimensionDefinitionLoose = CoordDimensionDefinition['name'] | C
+        }
+    }
+
+    // Make sure the first extra dim is 'value'.
+    const generateCoord = opt.generateCoord;
+    let generateCoordCount = opt.generateCoordCount;
+    const fromZero = generateCoordCount != null;
+    generateCoordCount = generateCoord ? (generateCoordCount || 1) : 0;
+    const extra = generateCoord || 'value';
+    let coordDimNameAutoIdx = 0;
+    let dataDimNameAutoIdx = 0;
+
+    // Set dim `name` and other `coordDim` and other props.
+    if (!omitUnusedDimensions) {
+        for (let resultDimIdx = 0; resultDimIdx < dimCount; resultDimIdx++) {
+            const resultItem = getResultItem(resultDimIdx);
+            const coordDim = resultItem.coordDim;
+
+            if (coordDim == null) {
+                const res = genName(
+                    extra, coordDimNameMap, coordDimNameAutoIdx, fromZero
+                );
+                coordDimNameAutoIdx = res.autoIdx;
+                resultItem.coordDim = res.name;
+                resultItem.coordDimIndex = 0;
+                // Series specified generateCoord is using out.
+                if (!generateCoord || generateCoordCount <= 0) {
+                    resultItem.isExtraCoord = true;
+                }
+                generateCoordCount--;
+            }
+
+            if (resultItem.name == null) {
+                const res = genName(
+                    resultItem.coordDim, dataDimNameMap, dataDimNameAutoIdx, false
+                );
+                resultItem.name = res.name;
+                dataDimNameAutoIdx = res.autoIdx;
+            }
+
+            if (resultItem.type == null
+                && (
+                    guessOrdinal(source, resultDimIdx) === BE_ORDINAL.Must
+                    // Consider the case:
+                    // {
+                    //    dataset: {source: [
+                    //        ['2001', 123],
+                    //        ['2002', 456],
+                    //        ...
+                    //        ['The others', 987],
+                    //    ]},
+                    //    series: {type: 'pie'}
+                    // }
+                    // The first colum should better be treated as a "ordinal" although it
+                    // might not able to be detected as an "ordinal" by `guessOrdinal`.
+                    || (resultItem.isExtraCoord
+                        && (resultItem.otherDims.itemName != null
+                            || resultItem.otherDims.seriesName != null
+                        )
+                    )
+                )
+            ) {
+                resultItem.type = 'ordinal';
+            }
+        }
+        return result;
+    }
+    else {
+        // Sort dimensions
+        const toSort = [];
+        for (let i = 0; i < indicesMap.length; i++) {
+            if (indicesMap[i] >= 0) {
+                toSort.push({ i, o: result[indicesMap[i]]});
+            }
+        }
+        toSort.sort((a, b) => a.i - b.i);
```

Comment:
Only in `SeriesData` will omit the unused dimensions. The `cloneAllDimensionInfo`  used in data transform has no relationship with it. It uses the `dimension define from the source.

Reply:
> Only in SeriesData will omit the unused dimensions. The cloneAllDimensionInfo used in data transform has no relationship with it. It uses the `dimension define from the source. 

Oh yes. Do not need to care `cloneAllDimensionInfo`.

Only the test case above should be fixed.

Reply:
Not sure why but seems this test case works as 5.1.2

Reply:
With the same option above, 

In 5.1.2: 
<img width="301" alt="Screen Shot 2021-08-03 at 4 14 20 PM" src="https://user-images.githubusercontent.com/1956569/127981913-1f307c09-a3a9-49db-820d-915625c86bb8.png">

But in the newest commit 
ae9226ecc01eb60e5662f34c8bcf275dbece17d8 [ae9226ecc]
July 27, 2021 at 5:16:25 PM GMT+8

<img width="235" alt="Screen Shot 2021-08-03 at 4 16 21 PM" src="https://user-images.githubusercontent.com/1956569/127982144-e472f27a-ac12-4d35-8936-c68eacf9c280.png">

Reply:
@100pah Understand. The rule of picking the last dimension as value is broken.

Reply:
@100pah Fixed it. The original `SeriesData#getDimensionInfo().index` is replaced by `SeriesData#getDimensionIndex`.
Which is always correct for reading `SeriesData#getRawValue`.

Currently the relationship between `SeriesData#dimensions` and `SeriesData#getStorageDimensions` is a bit confusing and I commented to explain it. But still not sure if we will use it wrong further.

Reply:
The index of `SeriesData['dimensions']` and `SeriesData['getDimension']` is not the dimensionIndex that users known, but the internal dimension index if dimension count > 30. 
But there are someplace still output that index to users:
+ in `CustomView`: `wrapEncodeDef(customSeries.getData())`), 
or use user input dimension index on `SeriesData['dimensions']`/`SeriesData['getDimension']`:
+ d
And `RadarView` has bug if use `encode` before this PR.
```js
```


    const userParams: CustomSeriesRenderItemParams = {
        // The life cycle of context: current round of rendering.
        // The global life cycle is probably not necessary, because
        // user can store global status by themselves.
        context: {},
        seriesId: customSeries.id,
        seriesName: customSeries.name,
        seriesIndex: customSeries.seriesIndex,
        coordSys: prepareResult.coordSys,
        dataInsideLength: data.count(),
        encode: wrapEncodeDef(customSeries.getData())
    } as CustomSeriesRenderItemParams;